### PR TITLE
Fix batching partition semantics for natural week/year/month.

### LINKED ISSUE #996: create batch partitions is not natural week/year/moth.
### Steps to reproduce the behavior (Required)
1. CREATE TABLE test2 (
    k1 DATE,
    k2 INT,
    k3 SMALLINT,
    v1 VARCHAR(2048),
    v2 DATETIME DEFAULT "2014-02-04 15:36:00"
)
ENGINE=olap
DUPLICATE KEY(k1, k2, k3)
PARTITION BY RANGE (k1) (
    START ("2020-12-24") END ("2021-10-03") EVERY (INTERVAL 1 MONTH)
)
DISTRIBUTED BY HASH(k2) BUCKETS 6
PROPERTIES (
    "replication_num" = "1",
    "dynamic_partition.enable" = "true",
    "dynamic_partition.time_unit" = "MONTH",
    "dynamic_partition.time_zone" = "Asia/Shanghai",
    "dynamic_partition.start" = "-2147483648",
    "dynamic_partition.end" = "1",
    "dynamic_partition.prefix" = "p",
    "dynamic_partition.buckets" = "6",
    "dynamic_partition.start_day_of_month" = "4",
    "in_memory" = "false",
    "storage_format" = "DEFAULT"
);

### Expected behavior (Required)
CREATE TABLE `test2` (
  `k1` date NULL COMMENT "",
  `k2` int(11) NULL COMMENT "",
  `k3` smallint(6) NULL COMMENT "",
  `v1` varchar(2048) NULL COMMENT "",
  `v2` datetime NULL DEFAULT "2014-02-04 15:36:00" COMMENT ""
) ENGINE=OLAP 
DUPLICATE KEY(`k1`, `k2`, `k3`)
COMMENT "OLAP"
PARTITION BY RANGE(`k1`)
(PARTITION p202012 VALUES [('2020-12-24'), ('2021-01-04')),
PARTITION p202101 VALUES [('2021-01-04'), ('2021-02-04')),
PARTITION p202102 VALUES [('2021-02-04'), ('2021-03-04')),
PARTITION p202103 VALUES [('2021-03-04'), ('2021-04-04')),
PARTITION p202104 VALUES [('2021-04-04'), ('2021-05-04')),
PARTITION p202105 VALUES [('2021-05-04'), ('2021-06-04')),
PARTITION p202106 VALUES [('2021-06-04'), ('2021-07-04')),
PARTITION p202107 VALUES [('2021-07-04'), ('2021-08-04')),
PARTITION p202108 VALUES [('2021-08-04'), ('2021-09-04')),
PARTITION p202109 VALUES [('2021-09-04'), ('2021-10-03')),
PARTITION p202110 VALUES [('2021-10-04'), ('2021-11-04')),
PARTITION p202111 VALUES [('2021-11-04'), ('2021-12-04')))
DISTRIBUTED BY HASH(`k2`) BUCKETS 6 
PROPERTIES (
"replication_num" = "1",
"dynamic_partition.enable" = "true",
"dynamic_partition.time_unit" = "MONTH",
"dynamic_partition.time_zone" = "Asia/Shanghai",
"dynamic_partition.start" = "-2147483648",
"dynamic_partition.end" = "1",
"dynamic_partition.prefix" = "p",
"dynamic_partition.buckets" = "6",
"dynamic_partition.start_day_of_month" = "4",
"in_memory" = "false",
"storage_format" = "DEFAULT"
);
### Real behavior (Required)
CREATE TABLE `test2` (
  `k1` date NULL COMMENT "",
  `k2` int(11) NULL COMMENT "",
  `k3` smallint(6) NULL COMMENT "",
  `v1` varchar(2048) NULL COMMENT "",
  `v2` datetime NULL DEFAULT "2014-02-04 15:36:00" COMMENT ""
) ENGINE=OLAP 
DUPLICATE KEY(`k1`, `k2`, `k3`)
COMMENT "OLAP"
PARTITION BY RANGE(`k1`)
(PARTITION p202012 VALUES [('2020-12-24'), ('2021-01-24')),
PARTITION p202101 VALUES [('2021-01-24'), ('2021-02-24')),
PARTITION p202102 VALUES [('2021-02-24'), ('2021-03-24')),
PARTITION p202103 VALUES [('2021-03-24'), ('2021-04-24')),
PARTITION p202104 VALUES [('2021-04-24'), ('2021-05-24')),
PARTITION p202105 VALUES [('2021-05-24'), ('2021-06-24')),
PARTITION p202106 VALUES [('2021-06-24'), ('2021-07-24')),
PARTITION p202107 VALUES [('2021-07-24'), ('2021-08-24')),
PARTITION p202108 VALUES [('2021-08-24'), ('2021-09-24')),
PARTITION p202109 VALUES [('2021-09-24'), ('2021-10-23')),
PARTITION p202110 VALUES [('2021-10-24'), ('2021-11-24'))
DISTRIBUTED BY HASH(`k2`) BUCKETS 6 
PROPERTIES (
"replication_num" = "1",
"dynamic_partition.enable" = "true",
"dynamic_partition.time_unit" = "MONTH",
"dynamic_partition.time_zone" = "Asia/Shanghai",
"dynamic_partition.start" = "-2147483648",
"dynamic_partition.end" = "1",
"dynamic_partition.prefix" = "p",
"dynamic_partition.buckets" = "6",
"dynamic_partition.start_day_of_month" = "4",
"in_memory" = "false",
"storage_format" = "DEFAULT"
);
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
1.19

### PR DESCRIPTION
**Pull request by Astralidea**
Fix #996 .
batch partitions semantics should match dynamic partitions.

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/analysis/MultiRangePartitionDesc.java [hunk 1]
```diff
@@ -142,18 +142,30 @@ public void setTimeUnit(String timeUnit) {
...
+                    DynamicPartitionUtil.checkStartDayOfWeek(dayOfWeekStr);
+                } catch (DdlException e) {
+                    throw new AnalysisException(e.getMessage());
+                }
+                dayOfWeek = Integer.parseInt(dayOfWeekStr);
+            }
+            if (properties.containsKey(DynamicPartitionProperty.START_DAY_OF_MONTH)) {
+                String dayOfMonthStr = properties.get(DynamicPartitionProperty.START_DAY_OF_MONTH);
+                try {
+                    DynamicPartitionUtil.checkStartDayOfWeek(dayOfMonthStr);
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/analysis/MultiRangePartitionDesc.java:159]: why check dayofweek？

File: fe/fe-core/src/main/java/com/starrocks/analysis/MultiRangePartitionDesc.java [hunk 2]
```diff
@@ -167,15 +179,27 @@ public void setTimeUnit(String timeUnit) {
...
+                        beginTime = finalEndTime;
+                    }
                     break;
                 case YEAR:
                     partitionName = DEFAULT_PREFIX + beginTime.toString(DateUtils.YEAR_FORMAT);
                     beginTime = beginTime.plusYears(timeInterval);
+                    beginTime = beginTime.withDayOfYear(1);
+                    if (beginTime.isAfter(finalEndTime)) {
+                        beginTime = finalEndTime;
+                    }
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/analysis/MultiRangePartitionDesc.java:202]: Why not do it after switch clause?

**Review by imay — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/analysis/MultiRangePartitionDesc.java [hunk 3]
```diff
@@ -142,18 +142,30 @@ public void setTimeUnit(String timeUnit) {
...
+                }
+                dayOfMonth = Integer.parseInt(dayOfMonthStr);
             }
-            dayOfWeek = Integer.parseInt(dayOfWeekStr);
         }
         WeekFields weekFields = WeekFields.of(DayOfWeek.of(dayOfWeek), 1);
-
-        while (beginTime.isBefore(endTime)) {
+        DateTime finalEndTime = new DateTime(endTime);
+        while (beginTime.isBefore(finalEndTime)) {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/analysis/MultiRangePartitionDesc.java:168]: Why not use endTime directly

File: fe/fe-core/src/main/java/com/starrocks/analysis/MultiRangePartitionDesc.java [hunk 4]
```diff
@@ -167,20 +179,26 @@ public void setTimeUnit(String timeUnit) {
...
                     break;
                 case MONTH:
                     partitionName = DEFAULT_PREFIX + beginTime.toString(DateUtils.MONTH_FORMAT);
+                    beginTime = beginTime.withDayOfMonth(dayOfMonth);
                     beginTime = beginTime.plusMonths(timeInterval);
                     break;
                 case YEAR:
                     partitionName = DEFAULT_PREFIX + beginTime.toString(DateUtils.YEAR_FORMAT);
                     beginTime = beginTime.plusYears(timeInterval);
+                    beginTime = beginTime.withDayOfYear(1);
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/analysis/MultiRangePartitionDesc.java:193]: Better to withDayOfYear before plusYears.

**Review by wyb — APPROVED** [review]

**Review by imay — APPROVED** [review]

**Pull request merged**